Source/GameMapEx.cpp

```diff
@@ -114,3 +114,29 @@ int GameMapEx::LoadMap(char *fileName)
 	
 	return func(p, 0, fileName);
 }
+
+void GameMapEx::CopyTileMap(int* tileMap, int xMin, int xMax, int yMin, int yMax)
+{
+	OP2Map *p = (OP2Map*)mapObj;
+	int **tileArray = (int**)(mapTileData);
+
+	int mapWidth = xMax - xMin;
+
+	for (int x = xMin; x < xMax; ++x)
+	{
+		for (int y = yMin; y < yMax; ++y)
+		{
+			int px = x - xMin;
+			int py = y - yMin;
+
+			int xLower = p->tileXMask & x;
+			int xUpper = xLower >> 5;
+			xLower &= 31;
+			xUpper <<= p->logTileHeight;
+			xUpper += y;
+			xUpper <<= 5;

```

Comment:
Might be worth extracting this calculation to an `inline`-able function.

Reply:
I was thinking a memcpy would be preferred, but I don't understand the tilemap format well enough to do it.

Reply:
I'm afraid a `memcpy` wouldn't work here. Particularly not for arbitrary alignments. The internal format of map data is not a simple 2D array. It's more like a 2D array sliced up into a 3D array. The end result likely has better cache friendliness for most common operations, though makes it more involved to copy a section out to a normal 2D array.

I guess a part of me is wondering why we would want to copy to a normal 2D array. This seems rather use case specific. It would seem more in the spirit of these library to access the data in place.